pkg/watch/watch.go

```diff
@@ -193,205 +199,145 @@ func (o *WatchClient) WatchAndPush(out io.Writer, parameters WatchParameters) er
-				// Also weirdly, fsnotify raises a RENAME event for deletion of files/folders with space in their name so even that should be handled here
-				if event.Op&fsnotify.Remove == fsnotify.Remove || event.Op&fsnotify.Rename == fsnotify.Rename {
-					// On remove/rename, stop watching the resource
-					if e := watcher.Remove(event.Name); e != nil {
-						klog.V(4).Infof("error removing watch for %s: %v", event.Name, e)
-					}
-					// Append the file to list of deleted files
-					// When a file/folder is deleted, it raises 2 events:
-					//	a. RENAME with event.Name empty
-					//	b. REMOVE with event.Name as file name
-					if !alreadyInChangedFiles && !matched && event.Name != "" {
-						deletedPaths = append(deletedPaths, event.Name)
-					}
-				} else {
-					// On other ops, recursively watch the resource (if applicable)
-					if e := addRecursiveWatch(watcher, event.Name, parameters.FileIgnores); e != nil && watchError == nil {
-						klog.V(4).Infof("Error occurred in addRecursiveWatch, setting watchError to %v", e)
-						watchError = e
-					}
-				}
-				changeLock.Unlock()
-			case watchErr := <-watcher.Errors:
-				changeLock.Lock()
-				watchError = fmt.Errorf("error watching filesystem for changes: %v", watchErr)
-				changeLock.Unlock()
-			}
-		}
-	}()
 	// adding watch on the root folder and the sub folders recursively
 	// so directory and the path in addRecursiveWatch() are the same
 	err = addRecursiveWatch(watcher, parameters.Path, parameters.FileIgnores)
 	if err != nil {
 		return fmt.Errorf("error watching source path %s: %v", parameters.Path, err)
 	}
 
+	printInfoMessage(out, parameters.Path)
+
+	var (
+		timer     *time.Timer
+		lastEvent fsnotify.Event
+	)
+	timer = time.NewTimer(time.Millisecond)
+	<-timer.C
 	// Only signal start of watch if invoker is interested
 	if parameters.StartChan != nil {
 		parameters.StartChan <- true
 	}
-
-	var ticker *time.Ticker
-	delay := time.Duration(parameters.PushDiffDelay) * time.Second
-
-	// don't create a ticker if delay is 0 as it will trigger panic
-	if delay != 0 {
-		ticker = time.NewTicker(delay)
-		defer ticker.Stop()
-	}
-	showWaitingMessage := true
-
-	hasFirstSuccessfulPushOccurred := false
-
-	// This for{} loop waits for filesystem changes that are signaled by the above goroutine;
-	// - 'dirty' is used by the goroutine to indicate that at least one change has occurred
 	for {
-		changeLock.Lock()
-		if watchError != nil {
-			klog.V(4).Infof("Ending watch for {} loop with error %v\n", watchError)
+		select {
+		case <-parameters.ExtChan:
+			watchError = ErrUserRequestedWatchExit
 			return watchError
-		}
-		if showWaitingMessage {
-			if parameters.EnvSpecificInfo != nil && parameters.EnvSpecificInfo.GetRunMode() == envinfo.Debug {
-				log.Finfof(out, "Component is running in debug mode\nPlease start port-forwarding in a different terminal.")
+		case event := <-watcher.Events:
+			lastEvent = event
+			timer.Reset(100 * time.Millisecond)
```

Comment:
I don't know, I still don't understand timer well. But I tried what you suggested.

Replacing:
```go
timer.Reset(100 * time.Millisecond)
``` 
with
```go
if !timer.Stop(){
    <-timer.C
}
timer.Reset(100 * time.Millisecond)
```
causes `odo dev` to not respond to a change in the file. That is, it doesn't trigger a push when we make a change in component directory.



What I think (my understanding of timers is still evolving) is happening is:
1. when there's an event, it gets stored in `lastEvent` and the timer is reset, which prevents the `case <-timer.C` from getting triggered.
2. this keeps happening a few times because a single file save triggers multiple fsnotify events.
3. finally, when all events triggered as a result of that single save are reported by fsnotify and overridden in `lastEvent` variable, the `case <- timer.C` gets triggered a 100ms after step 2 stops looping.